### PR TITLE
feat: add --fail-hook-affected-tests option to report skipped tests as failed

### LINKED ISSUE #4392: 🚀 Feature: Failures in before() and beforeAll() should cause all impacted tests to be reported as failed
### Description

every other testing suite that i've come across in my career (including, since it's a peer-ish library to `mocha`, [`jest`](https://github.com/facebook/jest)) will consistently report results for all tests on each run - more specifically,  it will show a test failure (or error, which is distinct from a "failure", depending on the framework) for each test in the suite(s) you're trying to execute, regardless of whether that failure occurred in a setup phase or in the test itself. `mocha`, however, appears to roll test setup failures that occur in `before`/`beforeEach` hooks into a single "failure", effectively swallowing counts of all the tests that were sidelined by the setup failure. this makes it rather frustrating to compare outputs between runs, understand the number of tests impacted by that setup failure at a glance, and can make reporting metrics from test runs somewhat meaningless due to the variance in the number of tests actually being reported on.

additionally, it seems that a failure in `beforeEach` will just circuit-break the execution of that setup function entirely, instead of retrying for subsequent tests, which honestly contravenes the name and intent of the function in the context of software testing conventions.

### Steps to Reproduce

- on any test suite with a `before` or `beforeEach` hook, introduce an intentional failure in the hook.
  - for your reference, i've thrown together a [simple demonstration](https://github.com/jmiller-airfox/mocha-simple-playground)
    - [with failing `before`](https://github.com/jmiller-airfox/mocha-simple-playground/tree/bad-behavior-before)
    - [with failing `beforeEach`](https://github.com/jmiller-airfox/mocha-simple-playground/tree/bad-behavior-beforeEach)
- run `npm test`
- `--reporter` options don't appear to have any affect on this

**Expected behavior:**
- a suite containing `N` tests (regardless of how they're nested in subsuites) should report on the outcome of `N` tests, even if there is a failure in a `before` or `beforeEach` hook.
- examples (modified from output of failing runs) of what i'd expect to see:
  - [`before`](https://gist.githubusercontent.com/jmiller-airfox/e22ff4e4fa23ce401ab5ebb2a9aae27f/raw/58b918239cba1742d9afd392eab003ea404bceb1/failure.before.proposal.txt)
  - [`beforeEach`](https://gist.githubusercontent.com/jmiller-airfox/2477ba5b12e1a6b2bf5aee20c4c7ec90/raw/50df8c58be20a451630e84e1a2b05321befed752/failure.beforeEach.proposal.txt)

**Actual behavior:** [What actually happens]
- a suite with a failure in a `before` or `beforeEach` hook will only report a single failure.
  - [`before`](https://gist.githubusercontent.com/jmiller-airfox/807a73692e1cf12556babec5235fc537/raw/e8e677679432a06465138ba386282a4d69b45793/failure.before.terminal.txt)
  - [`beforeEach`](https://gist.githubusercontent.com/jmiller-airfox/38ab7c8cf60ed5acc36dc6f234d7d352/raw/81718fa7adcff9018414c1a4f1cc84a1f4c9e157/failure.beforeEach.terminal.txt)

**Reproduces how often:** 100% - it seems to be by design..? so perhaps this should be a feature request instead?

### Versions

- `mocha --version`: `8.0.1`
- `node node_modules/.bin/mocha --version`: `8.0.1`
- `node --version`: `v12.18.1`
- Operating system
  - name and version: `osX Catalina 10.15.5`
  - architecture (32 or 64-bit): `x64`
- Shell (e.g., bash, zsh, PowerShell, cmd): `zsh 5.7.1 (x86_64-apple-darwin19.0)`
- browser and version: N/A
- Any third-party Mocha-related modules (and their versions): N/A
- Any code transpiler (e.g., TypeScript, CoffeeScript, Babel) being used (and its version): none

### PR DESCRIPTION
**Pull request by akshah123**
When `before()` or `beforeEach()` hooks fail, Mocha currently only reports the hook failure and silently skips affected tests. This makes it difficult to track test coverage and understand the full impact of hook failures.

This change introduces a new CLI option `--fail-hook-affected-tests` that, when enabled, reports all tests affected by hook failures as failed instead of silently skipping them.

Changes:
- Added --fail-hook-affected-tests boolean CLI option
- Implemented failAffectedTests() method in Runner to fail all affected tests
- Updated hook() method to call failAffectedTests when hooks fail
- Added failHookAffectedTests option to Mocha class
- Added integration tests for both before() and beforeEach() hook failures
- All affected tests now receive descriptive error messages indicating which hook caused them to be skipped

Fixes #4392

## PR Checklist

- [x] Addresses an existing open issue: fixes #4392
- [x] That issue was marked as [`status: accepting prs`](https://github.com/mochajs/mocha/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/mochajs/mocha/blob/main/.github/CONTRIBUTING.md) were taken

### CODE REVIEW

**Comment by mark-wiemer** [comment]
Hmm, I think we can ignore the CLA issue as the non-signer is AI :D

**Comment by mark-wiemer** [comment]
I'll review this in more detail sometime in the next ~7 days :)

File: lib/runner.js [hunk 1]
```diff
@@ -441,6 +442,47 @@ Runner.prototype.checkGlobals = function (test) {
...
+  var errorMessage =
+    'Test skipped due to failure in "' + hookTitle + '": ' + hookError.message;
+  var testError = new Error(errorMessage);
+  testError.stack = hookError.stack;
+
+  // Recursively fail all tests in this suite and its child suites
+  function failTestsInSuite(s) {
+    s.tests.forEach(function (test) {
+      // Only fail tests that haven't been executed yet
+      if (!test.state) {
```
Comment (mark-wiemer) [inline lib/runner.js:497]: Is `test.state` falsy for all tests that haven't been executed?

File: lib/runner.js [hunk 2]
```diff
@@ -441,6 +442,47 @@ Runner.prototype.checkGlobals = function (test) {
...
+        self.failures++;
+        self.emit(constants.EVENT_TEST_BEGIN, test);
+        self.emit(constants.EVENT_TEST_FAIL, test, testError);
+        self.emit(constants.EVENT_TEST_END, test);
+      }
+    });
+
+    s.suites.forEach(function (childSuite) {
+      failTestsInSuite(childSuite);
+    });
```
Comment (mark-wiemer) [inline lib/runner.js]:
```suggestion
    s.suites.forEach(failTestsInSuite);
```

File: lib/runner.js [hunk 3]
```diff
@@ -441,6 +442,47 @@ Runner.prototype.checkGlobals = function (test) {
...
+ * @param {string} hookTitle - The title of the failed hook
+ */
+Runner.prototype.failAffectedTests = function (suite, hookError, hookTitle) {
+  if (!this._opts.failHookAffectedTests) {
+    return;
+  }
+
+  var self = this;
+  var errorMessage =
+    'Test skipped due to failure in "' + hookTitle + '": ' + hookError.message;
```
Comment (mark-wiemer) [inline lib/runner.js]: Let's try to make this `errorMessage` more global so we don't have to write it multiple times. Maybe it should be a func that takes `hookTitle` and `hookErrorMsg` ?

File: test/integration/fixtures/hooks/before-each-hook-error-with-fail-affected.fixture.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+'use strict';
+
+describe('spec 1', function () {
```
Comment (mark-wiemer) [inline test/integration/fixtures/hooks/before-each-hook-error-with-fail-affected.fixture.js]:
```suggestion
describe('fails `beforeEach` hook', function () {
```

File: test/integration/fixtures/hooks/before-each-hook-error-with-fail-affected.fixture.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    throw new Error('before each hook error');
+  });
+  it('test 1', function () {
+    // This should be reported as failed due to beforeEach hook failure
+  });
+  it('test 2', function () {
+    // This should be reported as failed due to beforeEach hook failure
+  });
+});
+describe('spec 2', function () {
```
Comment (mark-wiemer) [inline test/integration/fixtures/hooks/before-each-hook-error-with-fail-affected.fixture.js]:
```suggestion
describe('passes normally', function () {
```

File: test/integration/fixtures/hooks/before-hook-error-with-fail-affected.fixture.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+'use strict';
+
+describe('spec 1', function () {
```
Comment (mark-wiemer) [inline test/integration/fixtures/hooks/before-hook-error-with-fail-affected.fixture.js]:
```suggestion
describe('fails `before` hook', function () {
```

File: test/integration/fixtures/hooks/before-each-hook-error-with-fail-affected.fixture.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+'use strict';
+
+describe('spec 1', function () {
+  beforeEach(function () {
+    throw new Error('before each hook error');
```
Comment (mark-wiemer) [inline test/integration/fixtures/hooks/before-each-hook-error-with-fail-affected.fixture.js]:
```suggestion
    throw new Error('error in `beforeEach` hook');
```

File: test/integration/fixtures/hooks/before-hook-error-with-fail-affected.fixture.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  before(function () {
+    throw new Error('before hook error');
```
Comment (mark-wiemer) [inline test/integration/fixtures/hooks/before-hook-error-with-fail-affected.fixture.js]:
```suggestion
    throw new Error('error in `before` hook');
```

File: test/integration/hook-err.spec.js [hunk 9]
```diff
@@ -267,6 +267,52 @@ describe("hook error handling", function () {
     });
   });
 
+  describe("--fail-hook-affected-tests", function () {
+    describe("before hook error", function () {
```
Comment (mark-wiemer) [inline test/integration/hook-err.spec.js]: I know this will dupe the beginning of each suite name, but really want to clarify these meta-names. We are not at a position prior to a hook error, we are referring to the hook named `before`.

```suggestion
    describe("error in `before` hook", function () {
```

File: test/integration/hook-err.spec.js [hunk 10]
```diff
@@ -267,6 +267,52 @@ describe("hook error handling", function () {
...
+              .and("to have failed test", "test 2")
+              .and("to have passed test count", 1)
+              .and("to have passed test", "test 3");
+            done();
+          },
+        );
+      });
+    });
+
+    describe("beforeEach hook error", function () {
```
Comment (mark-wiemer) [inline test/integration/hook-err.spec.js]:
```suggestion
    describe("error in `beforeEach` hook", function () {
```

**Review by mark-wiemer — CHANGES_REQUESTED** [review]
Overall looks OK, just requesting some renames for clarity. I'll check the coverage and make sure I can repro this one locally as it's a big change. Fortunately we should be able to include this in 12.0.0-beta-1 :)

File: lib/runner.js (hunk 1 above)
Comment (akshah123) [reply lib/runner.js:497]: That's what I see in `lib/runnable.js` file.  state is removed while resetting.  Once executed, they get one of these states:

STATE_PASSED (from lib/runnable.js)
STATE_FAILED
STATE_PENDING

**Review by mark-wiemer — APPROVED** [review]

**Comment by mark-wiemer** [comment]
Took me a minute to remember I needed to actively add the `--fail-hook-affected-tests` flag to see the new behavior :D was able to repro locally and the code does look good. Going to wait for @JoshuaKGoldberg to review, hopefully early next week.

File: lib/mocha.js [hunk 11]
```diff
@@ -847,6 +847,11 @@ Mocha.prototype.dryRun = function (dryRun) {
  * @return {Mocha} this
  * @chainable
  */
```
Comment (JoshuaKGoldberg) [inline lib/mocha.js:857]: [Docs] The JSDoc just above this function `failHookAffectedTests`, should stay with `failZero` next. `failHookAffectedTests` should get its own docs.

File: test/integration/hook-err.spec.js [hunk 12]
```diff
@@ -267,6 +267,52 @@ describe("hook error handling", function () {
     });
   });
```
Comment (JoshuaKGoldberg) [inline test/integration/hook-err.spec.js]: [Testing] It's possible for tests to throw something that's not an `Error`. It's not common, but within that the most common culprit's I've seen are either `null` or a `string`. Test error handling code tends to be brittle about this annoying edge case. Could you please add a test for each of those two types of non-`Error` exceptions?

File: lib/runner.js [hunk 13]
```diff
@@ -441,6 +442,58 @@ Runner.prototype.checkGlobals = function (test) {
...
+ * Create an error object for a test that was skipped due to a hook failure.
+ *
+ * @private
+ * @param {string} hookTitle - The title of the failed hook
+ * @param {Error} hookError - The error from the failed hook
+ * @returns {Error} The error object for the skipped test
+ */
+function createHookSkipError(hookTitle, hookError) {
+  var errorMessage =
+    'Test skipped due to failure in "' + hookTitle + '": ' + hookError.message;
```
Comment (JoshuaKGoldberg) [inline lib/runner.js]: > `hookError.message`

[Bug] It's possible for tests to throw something that's not an `Error`. This code doesn't handle the following cases well:

* `throw null`: the reported message is `Runnable failed with falsy or undefined exception.`
* `undefined`: the reported message is `undefined`

**Comment by akshah123** [comment]
@JoshuaKGoldberg Thanks for the thorough review! I've addressed both of your comments in the latest commit (c171cf11):

**Re: [Docs] JSDoc placement** - Fixed! The JSDoc now correctly documents `failHookAffectedTests` and has been separated from `failZero`'s documentation. Each function now has its own properly positioned JSDoc comment.

**Re: [Bug] Non-Error handling** - Great catch! The `createHookSkipError` function now properly handles:
- Falsy values (null/undefined) using `createInvalidExceptionError`
- Non-Error objects (strings, numbers, etc.) using `thrown2Error`

This follows the existing patterns from `runner.js:1112` and `runner.js:537`. I've also added comprehensive test coverage for these edge cases with new fixture files testing null, undefined, string, and number throws in both `before` and `beforeEach` hooks.

All tests are passing, including the 2 new integration tests specifically for non-Error handling. ✅

File: lib/runner.js [hunk 14]
```diff
@@ -441,6 +442,71 @@ Runner.prototype.checkGlobals = function (test) {
...
+ * @private
+ * @param {string} hookTitle - The title of the failed hook
+ * @param {*} hookError - The error from the failed hook (may not be an Error object)
+ * @returns {Error} The error object for the skipped test
+ */
+function createHookSkipError(hookTitle, hookError) {
+  // Handle falsy or undefined exceptions
+  if (!hookError) {
+    hookError = createInvalidExceptionError(
+      'Hook "' + hookTitle + '" failed with falsy or undefined exception.',
```
Comment (JoshuaKGoldberg) [inline lib/runner.js]: [Feature] It's sometimes useful for folks to know if it's specifically `null`, `undefined`, `0`, etc.:

```suggestion
      'Hook "' + hookTitle + '" failed with a '" + hookError + "' exception.',
```

**Comment by mark-wiemer** [comment]
@akshah123 looks like your latest commit broke some stuff, please review the failed checks.

@JoshuaKGoldberg happy to port this back to v11, created #5580

**Comment by akshah123** [comment]
Sorry about that @mark-wiemer . I didn't verify syntax issues with suggested changes.  It should be fixed now.

File: lib/runner.js [hunk 15]
```diff
@@ -441,6 +442,71 @@ Runner.prototype.checkGlobals = function (test) {
...
+ * @private
+ * @param {string} hookTitle - The title of the failed hook
+ * @param {*} hookError - The error from the failed hook (may not be an Error object)
+ * @returns {Error} The error object for the skipped test
+ */
+function createHookSkipError(hookTitle, hookError) {
+  // Handle falsy or undefined exceptions
+  if (!hookError) {
+    hookError = createInvalidExceptionError(
+      'Hook "' + hookTitle + '" failed with a "' + hookError + '" exception.',
```
Comment (mark-wiemer) [inline lib/runner.js]:
```suggestion
      'Hook "' + hookTitle + '" failed with exception: ' + hookError,
```

File: lib/runner.js [hunk 16]
```diff
@@ -441,6 +442,71 @@ Runner.prototype.checkGlobals = function (test) {
...
+      hookError,
+    );
+  }
+  // Convert non-Error objects to Error
+  else if (!isError(hookError)) {
+    hookError = thrown2Error(hookError);
+  }
+
+  var errorMessage =
+    'Test skipped due to failure in "' + hookTitle + '": ' + hookError.message;
```
Comment (mark-wiemer) [inline lib/runner.js]:
```suggestion
    'Test skipped due to failure in hook "' + hookTitle + '": ' + hookError.message;
```

File: lib/runner.js [hunk 17]
```diff
@@ -441,6 +442,71 @@ Runner.prototype.checkGlobals = function (test) {
   }
 };
 
+/**
+ * Create an error object for a test that was skipped due to a hook failure.
+ * Handles cases where the thrown value is not an Error object.
```
Comment (mark-wiemer) [inline lib/runner.js]:
```suggestion
```

This is already covered by the param type and description :)

File: lib/runner.js (hunk 1 above)
Comment (mark-wiemer) [reply lib/runner.js:497]: I reviewed this as well, matches my understanding :)

**Comment by mark-wiemer** [comment]
Still LGTM to me, great work :) running tests now to make sure nothing weird changed, then I plan to apply my suggestions and run tests once more. @JoshuaKGoldberg , OK to merge this into 12.0 or should we wait until 12.1? Obviously the flag is still disabled by default so I don't foresee issues and would love to merge this into 12.0 :)

**Comment by JoshuaKGoldberg** [comment]
If we merge it into 12, I think we should probably also back-port it to 11. 👍 from me!

**Comment by mark-wiemer** [comment]
Yep, we already have  https://github.com/mochajs/mocha/issues/5580 :)

**Comment by JoshuaKGoldberg** [comment]
_heck yes_
